cdap-app-fabric/src/main/java/co/cask/cdap/metadata/profile/ProfileMetadataMessageProcessor.java

```diff
@@ -0,0 +1,250 @@
+          updateProgramProfileMetadata(programId);
+        }
+        break;
+      case SCHEDULE:
+        ScheduleId scheduleId = (ScheduleId) entityId;
+        // make sure the schedule exists before updating
+        try {
+          ProgramSchedule schedule = scheduleDataset.getSchedule(scheduleId);
+          updateScheduleProfileMetadata(schedule, getResolvedProfileId(schedule.getProgramId()));
+        } catch (NotFoundException e) {
+          LOG.warn("Schedule {} is not found, so its profile metadata will not get updated. " +
+                     "Ignoring the message {}", scheduleId, message);
+          return;
+        }
+
+        break;
+      default:
+        // this should not happen
+        LOG.warn("Type of the entity id {} cannot be used to update profile metadata. " +
+                   "Ignoring the message {}", entityId, message);
+    }
+  }
+
+  private void removeProfileMetadata(Set<EntityId> entityIds) {
+    for (EntityId entityId : entityIds) {
+      // now profile metadata will only be on the program or schedule
+      if (entityId instanceof ProgramId || entityId instanceof ScheduleId) {
+        metadataDataset.removeProperties((NamespacedEntityId) entityId, PROFILE_METADATA_KEY);
+      }
+    }
+  }
+
+  private void updateAppProfileMetadata(ApplicationId applicationId, ApplicationSpecification appSpec) {
+    for (String name : appSpec.getWorkflows().keySet()) {
+      WorkflowId programId = applicationId.workflow(name);
+      updateProgramProfileMetadata(programId);
+    }
+  }
+
+  private void updateProgramProfileMetadata(ProgramId programId) {
+    ProfileId profileId = getResolvedProfileId(programId);
+    setProfileMetadata(programId, profileId);
+
+    for (ProgramSchedule schedule : scheduleDataset.listSchedules(programId)) {
+      updateScheduleProfileMetadata(schedule, profileId);
+    }
+  }
+
+  private void updateScheduleProfileMetadata(ProgramSchedule schedule, @Nullable ProfileId profileId) {
+    ScheduleId scheduleId = schedule.getScheduleId();
+    Optional<ProfileId> scheduleProfileId =
+      SystemArguments.getProfileIdFromArgs(scheduleId.getNamespaceId(), schedule.getProperties());
+    profileId = scheduleProfileId.isPresent() ? scheduleProfileId.get() : profileId;
+    setProfileMetadata(scheduleId, profileId);
+  }
+
+  private void setProfileMetadata(NamespacedEntityId entityId, @Nullable ProfileId profileId) {
+    // if we are able to get profile from preferences or schedule properties, use it
+    // otherwise default profile will be used
+    metadataDataset.setProperty(entityId, PROFILE_METADATA_KEY,
+                                profileId == null ? ProfileId.DEFAULT.toString() : profileId.toString());
+  }
+
+  // TODO: CDAP-13579 consider preference key starts with [scope].[name].system.profile.name
+  @Nullable
+  private ProfileId getResolvedProfileId(EntityId entityId) {
+    if (entityId instanceof InstanceId) {
```

Comment:
you should use the entity type instead of instanceof

Reply:
Fixed